### PR TITLE
fix(model): convert legacy snake_case attachments on decode

### PR DESCRIPTION
**Pull request by evelynchang5525**
## Summary

Pre-migration Cassandra rows store attachments with `snake_case` JSON keys (`title_link`, `image_url`, etc.), but `cassandra.Attachment` declares `camelCase` tags. Since `encoding/json` field matching is case-insensitive but not underscore-insensitive, legacy blobs silently decode into incomplete attachments with no `id`, `titleLink`, or `fileType` — rendering and downloads fail.

This change detects legacy blobs at read time in `DecodeAttachments` and converts them to the current shape, rewriting download URLs onto `GET /api/v1/file-upload/:fileId/:fileName`.

## Key Changes

- **`pkg/model/cassandra/attachment_legacy.go`** (new): Pure conversion logic
  - `legacyAttachment` struct with `snake_case` tags
  - `normalizeAttachment()` — detects and converts legacy blobs in place
  - `convertLegacy()` — builds current-shape `Attachment` from legacy data
  - Helper functions: `legacyURLPath()`, `legacyDownloadURL()`, `legacyFileID()`, `legacyFileName()`, `fileTypeFromName()`
  - Handles absolute URLs, unrecognized path layouts, missing fields with sensible fallbacks

- **`pkg/model/cassandra/attachment.go`** (modified): Wire conversion into the decode loop
  - Call `normalizeAttachment()` after decoding each blob
  - Update doc comment to note legacy conversion

- **`pkg/model/cassandra/attachment_legacy_test.go`** (new): Comprehensive test coverage
  - Canonical issue #374 payload conversion
  - Edge cases: absolute URLs, already-prefixed paths, missing title/type/media-type, degenerate blobs
  - Current-format passthrough (byte-for-byte round-trip)
  - Mixed columns (legacy + current + malformed)
  - Idempotence (re-encode and decode again)

- **`docs/client-api.md`** (modified): Document server-side normalization
  - Note that legacy rows are converted transparently
  - Clarify that clients always receive the current schema

## Implementation Details

**Detection rule:** A blob is legacy if it decodes with empty `id` AND empty `titleLink`. This ensures current-format attachments (which always carry both) pass through untouched, and future `Attachment` fields are forward-compatible.

**Conversion:** Extract file ID from the path (segment after `file-upload`, fallback to second-to-last), rewrite URL to `api/v1/file-upload/{id}/{fileName}`, derive MIME type from declared media field or file extension, drop legacy image geometry and thumbnails by design.

**Data safety:** Cassandra is never rewritten — pure read-time function, idempotent, rollback-safe. Message-worker's quoted-parent snapshots are re-encoded in the current format (self-healing, not a regression).

**Scope:** Fixes all four read paths (`history-service`, `broadcast-worker`, `pkg/searchindex`, `message-gatekeeper`) with one change. No upload-service or ES backfill changes.

Refs #374

https://claude.ai/code/session_0177MP4zM8cwV4fNxaQukqTJ

## Summary by CodeRabbit

* **New Features**
  * Legacy attachments are now automatically converted to the current format when retrieved.
  * Converted attachments include consistent file links, IDs, titles, and file-type information.
  * Existing current-format attachments remain unchanged.

* **Bug Fixes**
  * Malformed or incomplete legacy attachments are safely skipped without disrupting other attachments.

* **Documentation**
  * Added API documentation and design details for legacy attachment handling.

### CODE REVIEW

File: docs/client-api.md [hunk 1]
```diff
@@ -884,6 +884,12 @@ Render-ready descriptor for an uploaded file. Returned by the upload endpoint
 into `msg.send` (§4), and returned decoded as objects in message payloads. Media
 fields are present only for the matching MIME family.
 
+Attachments stored in the pre-migration format are converted server-side on
+every read, so clients always receive the schema below. A converted attachment
+carries `id`, `title`, `type`, `titleLink`, `titleLinkDownload` and `fileType`
+(plus `description` when present); its `titleLink` points at
+`api/v1/file-upload/{fileId}/{fileName}`, and the media fields are absent.
```
[the remark this answers is not in the document]
Comment (evelynchang5525) [reply docs/client-api.md]: Confirmed and fixed in 985faa9 — I took the second option you offered (default the conversion to `true`) rather than documenting the preserved flag.

Rationale for defaulting rather than preserving: every converted `titleLink` is rewritten onto `api/v1/file-upload/{fileId}/{fileName}`, which resolves to `upload-service`'s `HandleDownloadMinioS3File`. That handler only ever responds with `Content-Disposition: attachment`, so `true` is factually correct for every converted attachment, matches what `buildAttachment` sets on new uploads, and keeps the documented "Always `true`" invariant intact instead of weakening it for one class of attachment.

The unused `TitleLinkDownload` field was dropped from `legacyAttachment`, and `TestDecodeAttachments_LegacyAlwaysDownloadable` covers both the absent and explicitly-`false` legacy cases.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: pkg/model/cassandra/attachment_legacy.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+	Description       string `json:"description"`
+	TitleLink         string `json:"title_link"`
+	TitleLinkDownload bool   `json:"title_link_download"`
+
+	ImageURL  string `json:"image_url"`
+	ImageType string `json:"image_type"`
+	AudioURL  string `json:"audio_url"`
+	AudioType string `json:"audio_type"`
+	VideoURL  string `json:"video_url"`
+	VideoType string `json:"video_type"`
```
[the remark this answers is not in the document]
Comment (evelynchang5525) [reply pkg/model/cassandra/attachment_legacy.go:34]: Not making this change — I don't think the guideline applies here.

`legacyAttachment` isn't a model struct. It's an unexported, decode-only helper that parses one JSON blob out of the Cassandra `attachments LIST<BLOB>` column. It is never marshalled, never persisted, and never touches MongoDB — the only operation performed on it is `json.Unmarshal`, after which it is converted to `Attachment` and discarded.

The package convention backs this up: there are zero `bson` tags anywhere in `pkg/model/cassandra` (all 10 files). That includes the exported `Attachment` type this struct converts into, which is `json`-only, and `cassandra.Message`, which uses `cql` tags. Where these types do reach MongoDB, the `bson` tag is applied at the embedding site in `pkg/model` — e.g. `pkg/model/message.go:24` tags the `*cassandra.Card` field, not the `Card` type's own fields.

Adding `bson` tags here would make `legacyAttachment` the only bson-tagged struct in the package, and inconsistent with the very type it converts into, while changing no behaviour.

Leaving this thread open in case @evelynchang5525 disagrees.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by mliu33 — APPROVED** [review]
Outstanding work, thanks! One thing just double check. I remember we already support legacy downloadLink path in upload-service, why do we need to rewrite it ?

**Comment by evelynchang5525** [comment]
Thanks for the review — and you're right that the legacy download endpoint already exists. That endpoint is exactly what the rewrite targets; what's missing isn't the handler, it's the `/api/v1` prefix on the *stored* path.

**What upload-service serves** (`upload-service/routes.go:15`), inside the `/api/v1` group:

```
GET /api/v1/file-upload/:fileId/:fileName   →  HandleDownloadMinioS3File
```

**What the legacy Cassandra blob stores:**

```json
"title_link": "/file-upload/xh3e4jnJDhEvEy7rk/image%20(2).png"
```

Leading slash, no `api/v1`. There is no route registered at bare `/file-upload/...` — `routes.go` only registers `/healthz`, the `/api/v1` group, and the `/api/v3` protected-image route.

The frontend then does `mediaUrl(baseUrl, att.titleLink)` (`chat-frontend/src/lib/attachment.js:65`), which strips leading slashes and joins onto the gateway base:

```js
return `${baseUrl.replace(/\/+$/, '')}/${relativePath.replace(/^\/+/, '')}`
```

So the raw legacy value resolves to `{gateway}/file-upload/{id}/{name}` → 404, while the rewritten `api/v1/file-upload/{id}/{name}` lands on your existing handler. It's a prefix fix, not new endpoint support.

**One caveat worth your judgement:** I can only see this repo. If your API gateway or ingress already rewrites `/file-upload/*` → `/api/v1/file-upload/*` outside this codebase, then the URL rewrite is redundant — harmless, since it's idempotent and `api/v1/...` is the documented current form, but redundant. If that's what you were remembering, say so and I'll drop the URL rewrite.

Worth noting either way: the URL is only one of three things the legacy blob is missing. It also has no `id` and no `fileType` at all, and its `snake_case` keys mean `encoding/json` silently drops `title_link`, `image_type` and the rest — a legacy attachment currently decodes to just `{title, type}`. So the conversion is needed regardless of how the URL resolves.

---
_Generated by [Claude Code](https://claude.ai/code)_
